### ISSUE TITLE
t.w.template does not always properly escape comments

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5275]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)| @bitglue reported|
|-|-|
|Trac ID|trac#5275|
|Type|defect|
|Created|2011-09-26 17:43:15Z|
|Branch|https://github.com/twisted/twisted/tree/template-comment-fix-5275|

```
>>> from twisted.web.template import flattenString, Comment
>>> from twisted.python.util import println
>>> flattenString(None, Comment('the quick---brown fox')).addCallback(println)
<!--the quick- --brown fox-->
<Deferred at 0x1020d6a28 current result: None>
```

Which of course is malformed HTML, because HTML is SGML, and comments in SGML begin and end with '--', and are between '<!' and '>'. Or something like that.

What this generates is a comment declaration with one comment, 'the quick- ', and some illegal characters 'brown fox--'.

http://htmlhelp.com/reference/wilbur/misc/comment.html

Attachments:

* [template-comments-5275.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/451/451e7470b4f5da501cd95543e8a874c4fb1aa542/39c1449b406e0b1ffaa4c2c9144b90783281e9d0.patch) (2065 bytes) - added by indigo on 2011-09-26 18:13:56Z - proposed fix
* [template-comments-5275.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/451/451e7470b4f5da501cd95543e8a874c4fb1aa542/ce4df89315ef0c5e13f42e5ed45f6692179fa1ac.patch) (2342 bytes) - added by indigo on 2011-09-26 18:18:03Z - 
* [template-comments-5275.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/451/451e7470b4f5da501cd95543e8a874c4fb1aa542/7f03e3e4645be4f12bcee6fd5cbda030a43ab58d.patch) (2654 bytes) - added by indigo on 2011-09-26 18:37:31Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5275 5275
type__defect defect
reporter__indigo indigo
priority__normal normal
milestone__ 
branch__branches_template_comment_fix_5275 branches/template-comment-fix-5275
branch_author__indigo indigo
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1317058995000000 1317058995000000
changetime__1317242625000000 1317242625000000
version__None None
owner__exarkun exarkun
cc__indigo
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_3"></a>|@bitglue commented|
|-|-|

I couldn't think of an elegant way to fix this without changing the way comments are serialized sometimes. Not sure anyone cares.

Also, the proposed patch escapes things which strictly need not be escaped, at least by SGML syntax rules. My reasoning was that it's better to be conservative and generate something that won't confuse a browser's parser and will be recognized as a comment than it is to follow SGML to the letter. There are plenty of valid SGML comments that confuse browsers anyway, so this seems reasonable.

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_4"></a>|@bitglue commented|
|-|-|

Seriously, comment syntax is complicated. Did you know they can't end in '-'? I didn't. http://www.w3.org/TR/REC-xml/#sec-comments

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [32674]) Branching to 'template-comment-fix-5275'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [32675]) Apply template-comments-5275.3.patch 

refs [#5275](https://github.com/twisted/twisted/issues/5275)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks a lot!  Some simple coding standard points:

  1. `test_CommentEscaping` should be `test_commentEscaping`.
  1. Test method docstrings should not start with "Test ...".
  1. Two blank lines between methods.

Some others:

  4. It might be nice if when this test failed it reported what kind of input it failed on and what bogus output it got.
  5. [ReleaseProcess#Newsfiles News fragments] are handy.  Committers will often write them for you, but it's nice when you write them. :)  This is a nice bug fix, it should definitely have a nice news fragment.

These points are pretty minor.  I'll address them and merge, if everything looks good on buildbot.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [32676]) Address some review feedback for parts of the new test

refs [#5275](https://github.com/twisted/twisted/issues/5275)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [32677]) news fragment

refs [#5275](https://github.com/twisted/twisted/issues/5275)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [32678]) Merge template-comment-fix-5275 

Author: indigo
Reviewer: exarkun
Fixes: [#5275](https://github.com/twisted/twisted/issues/5275)

Add comment value mangling to `twisted.web.template` so that comments are not
emitted which will break an SGML-ish or XML parser on the receiving end.  Since
there is no real escaping rule for comments in an HTML document, the value is
mangled irreversibly.
